[Storage] Use @azure/core-http in @azure/storage-* libraries to support @azure/identity

daviwil opened the pull request:
This PR updates the `@azure/storage-*` libraries to use the new `@azure/core-http` and its `TokenCredential` interface so that the `@azure/identity` library's credential implementations can be used with it.  The largest part of the changes is switching from `@azure/ms-rest-js` to `@azure/core-http` and the changes to the generated code from using an updated `autorest.typescript`.

The real meat of the changes is in the `*Client` classes like `QueueClient` where I add an additional `TokenCredential` type option for the `credential` argument and then set up the pipeline differently in `Pipeline.ts` depending on which credential has been passed.

**Questions for Reviewers**

- Should I split up the changes to the storage libraries into 3 separate PRs so that it doesn't get too crazy?
- Should I add a new source file under the `samples/` folder strictly for showing how to use `@azure/identity` with this library?

**Tasks**

- [x] Rename storage libraries' `TokenCredential` to `RawTokenCredential`
- [x] Add code changes for `storage-blob` and `storage-file`
- [x] Use `@azure/identity` in samples for all 3 libraries
- [x] Add tests to verify that the `Pipeline` uses a `BearerTokenAuthenticationPolicy` when a `TokenCredential` is passed
- [x] Remove type hints for model types that are mistakenly compared to `ServiceCallback<void>`
- [x] Update `BreakingChange.md` if necessary to mention spec change (https://github.com/Azure/azure-sdk-for-js/pull/3853#discussion_r294669791)

jeremymeng commented:
>     * Should I add a new source file under the `samples/` folder strictly for showing how to use `@azure/identity` with this library?

That would be valuable!

jeremymeng commented on sdk/storage/storage-queue/src/QueueClient.ts:
```diff
@@ -286,7 +297,7 @@ export class QueueClient extends StorageClient {
       ...options,
       abortSignal: aborter
-    });
+    } as Models.QueueCreateOptionalParams);
```
Interesting, I'd thought that this should be compatible with `Models.QueueCreateOptionalParams`. Curious why the cast is needed?

daviwil replied:
I was surprised about this too.  For some reason the TypeScript compiler is comparing the object sent as parameter to the `create()` method to the `ServiceCallback<void>` type instead of `Models.QueueCreateOptionalParams`.  I'd like to figure out the real reason why this is happening so that I can take the type hint back out.  I'll add this to my TODO list.

daviwil replied:
Looking at this a bit more closely, it seems that the `metadata` field of the options type has been changed from an object/dictionary to a string:

https://github.com/Azure/azure-sdk-for-js/pull/3853/files#diff-33e46db3382822f99e85f41eb27fd333L570

Trying to track down why the code generator would change this now.

jeremymeng replied:
Looks like it has a `string` type in the swagger json? I wonder how the previous code was generated to have a dictionary @XiaoningLiu.

daviwil replied:
I noticed that there's a `#/parameters/Metadata` and a `#/definitions/Metadata` definition where the former has the type set to `string` and the latter is `object` (which seems correct).  Either there's a bug in the Swagger spec or something in the code generator has changed such that it picks up the wrong type.  I'll look into the code generation angle to see if something got broken there.

XiaoningLiu replied:
@daviwil Which swagger you are using for the generation? Metadata should definitely be an `object`.

daviwil replied:
Turns out that this was a [bug in `@microsoft.azure/oai2-to-oai3`](https://github.com/Azure/perks/pull/58), I've regenerated the `StorageClient`s after this fix and the correct `metadata` type has now been restored.

jeremymeng commented on sdk/storage/storage-queue/src/generated/lib/operations/service.ts:
```diff
@@ -115,25 +115,53 @@ export class Service {
+   * @param [options] The optional parameters
+   * @returns Promise<Models.ServiceListQueuesSegmentResponse>
+   */
+  listQueuesSegmentNext(nextPageLink: string, options?: Models.ServiceListQueuesSegmentNextOptionalParams): Promise<Models.ServiceListQueuesSegmentResponse>;
```
Previously there's only one `listQueueSegment()` method.  Now we have both `listQueuesSegment()` and `listQueuesSegmentNext()` I wonder whether we there are any changes that we need to make in how we list queues.

daviwil replied:
I don't think any changes are necessary.  The reason why these showed up now is because I updated `autorest.typescript` to include methods for operations with results that support paging when the `--azure-arm` parameter *isn't* used.  These methods can be used when paging is desired but I think they can be ignored otherwise.

jeremymeng replied:
We do need paging: (https://github.com/Azure/azure-sdk-for-js/blob/feature/storage/sdk/storage/storage-queue/src/QueueServiceClient.ts#L289-L298)

jeremymeng replied:
Looks like `listQueuesSegment()` didn't change so it should still work. I guess `listQueuesSegmentNext()` is new addition that takes the marker - `nextPageLink` explicitly.

daviwil replied:
Yep, that's correct.  Interesting that results can be paged without these additions, I wonder why the `keyvault-*` libraries needed the explicit `*Next()` methods...

jeremymeng commented on sdk/storage/storage-queue/src/MessagesClient.ts:
```diff
@@ -180,12 +184,16 @@ export class MessagesClient extends StorageClient {
    *                     append a SAS if using AnonymousCredential, such as
    *                     "https://myaccount.queue.core.windows.net/myqueue/messages?sasString".
-   * @param {Credential} credential Such as AnonymousCredential, SharedKeyCredential or TokenCredential.
+   * @param {Credential | CoreHttpTokenCredential } credential Such as AnonymousCredential, SharedKeyCredential or TokenCredential.
```
There are `corehttp.TokenCredential` and `CoreHttpTokenCredential`. Better to be consistent.

daviwil replied:
I agree, I was also thinking that while making this change.  I might standardize on `CoreHttp*` instead of the `coreHttp.` approach because it occurs more often in the changes I made.

XiaoningLiu replied:
I'm worried about the renaming to `CoreHttpTokenCredential`.

Renaming is private in storage SDK. Customers won't know anything about `CoreHttpTokenCredential`. He only can get `TokenCredential` from identity or core-http packages.

XiaoningLiu replied:
How about this?

1. Rename Storage SDK's `TokenCredential` class to another name like `RawTokenCredential` or `SimpleTokenCredential`.
 Just some other name different with interface from core-http `TokenCredential`. This is the root conflict we need to resolve.

2. Update Storage SDK's `RawTokenCredential` class to implement `TokenCredential` interface provided by core-http.

3. In constructors of `QueueClient` and other Clients, make credential parameter to accept `Credential | TokenCredential`

```javascript
  constructor(
    url: string,
    credential?: Credential | TokenCredential,
    options?: NewPipelineOptions
  );
```

4. Update all documents and tell customers he can get a `TokenCredential` using identity package or `RawTokenCredential` provided in storage SDK.

In this way, we don't need the `CoreHttpTokenCredential` renaming. No confusing, and all TokenCredential will be aligned to have same definition.

daviwil replied:
I've renamed `TokenCredential` to `RawTokenCredential` as requested and updated references of `CoreHttpTokenCredential` to `TokenCredential`.

Regarding the union of `Credential | TokenCredential`,  I don't know if there's a way around that just yet.  Currently `@azure/core-http`'s `TokenCredential` interface is meant for authentication against Azure Active Directory and using the resulting Bearer token for authentication.  If we expand to other credential types in the future it's possible that we could deprecate `Credential` in favor of `TokenCredential`, but I'm not sure whether that is ideal.

jeremymeng commented:
/cc @XiaoningLiu I think this is a good start for enabling azure-identity. We can discuss better integration with current credentials in Storage but that may need to happen after this preview if it's taking longer to implement.

jeremymeng commented on sdk/storage/storage-queue/src/generated/lib/models/index.ts:
```diff
@@ -1,1372 +1,1051 @@
+   * response body.
    */
-  include?: ListQueuesIncludeType;
+  include?: ListQueuesIncludeType[];
```
This looks breaking too.  I wonder whether the original generated code is wrong, or intentionally changed to a singleton, as the only valid type here is `'metadata'`.  @XiaoningLiu do you remember anything?

daviwil replied:
Looks like it's represented as `array` in the [OpenAPI json](https://msazure.visualstudio.com/One/_git/Storage-XStore?path=%2Fsrc%2FXFE%2FOpenApi%2FMicrosoft.QueueStorage%2Fpreview%2F2018-03-28%2Fqueue.json&version=GBmaster&line=1711&lineStyle=plain&lineEnd=1712&lineStartColumn=1&lineEndColumn=1):

```json
    "ListQueuesInclude": {
      "name": "include", 
      "in": "query", 
      "required": false, 
      "type": "array",
      "collectionFormat": "csv",
      "items": {
      "type": "string", 
      "enum": [ 
        "metadata" 
      ], 
      "x-ms-enum": { 
        "name": "ListQueuesIncludeType", 
        "modelAsString": false 
        }
      },
      "x-ms-parameter-location": "method",
      "description": "Include this parameter to specify that the queues's metadata be returned as part of the response body."
    },
```

XiaoningLiu replied:
I believe this is a swagger breaking change introduced in https://msazure.visualstudio.com/One/_git/Storage-XStore/commit/76530600659f4d591d65266c74b2d96c785c3015?_a=compare&path=%2Fsrc%2FXFE%2FOpenApi%2FMicrosoft.QueueStorage%2Fpreview%2F2018-03-28%2Fqueue.json 

We can take this change. But please update BreakingChange.md for this.

XiaoningLiu commented:
Implementation should not take long, but the discussion may.. 
At least, in preview, let's figure out an clear naming strategy for storage build-in credential and identity credentials to avoid confusing to customers. 

BTW, in [https://github.com/azure/azure-sdk-for-js#authentication](https://github.com/azure/azure-sdk-for-js#authentication) , it recommends to use @azure/ms-rest-nodeauth and @azure/ms-rest-browserauth for authentication. What's the relationship between them with @azure/identity?

---
In reply to: [502846562](https://github.com/Azure/azure-sdk-for-js/pull/3853#issuecomment-502846562) [](ancestors = 502846562)

XiaoningLiu commented:
>     "build:autorest": "autorest ./swagger/README.md --typescript --use=@microsoft.azure/autorest.typescript@2.1.3",

@David Wilson For core-http change, do we need a newer autorest.typescript version? If so, please update the autorest.typescript version here.

---
Refers to: sdk/storage/storage-queue/package.json:21 in bcf748f. [](commit_id = bcf748f83ecab32d0bdfc09e19100dff390e7151, deletion_comment = False)

XiaoningLiu commented:
Is there any autorest generator configuration change? If so, please update swagger/readmd.md, and any configurations defined in package.json `build:autorest`

daviwil commented:
@XiaoningLiu Thanks for the great feedback!  I've made some of the changes you requested and also added `@azure/core-http` integration for `@azure/storage-file` and `@azure/storage-blob`, please take a look.

Regarding the AutoRest changes, I'm currently working out of the [`azure-core` branch](https://github.com/Azure/autorest.typescript/tree/azure-core) of `autorest.typescript` and unfortunately I don't have a published update yet (we're waiting to settle on the new `@azure/core-http` library so that we can deprecate `@azure/ms-rest-js` before doing this).  

No configuration changes are needed for the `swagger/README.md` files, it all works exactly the same as it did before.  I've been generating the new `StorageClients` using the specs in the repo you linked to before on msazure.visualstudio.com.

XiaoningLiu commented on sdk/storage/storage-blob/src/credentials/RawTokenCredential.ts:
```diff
@@ -26,39 +23,36 @@ import { TokenCredentialPolicy } from "../policies/TokenCredentialPolicy";
    */
-  public create(nextPolicy: RequestPolicy, options: RequestPolicyOptions): TokenCredentialPolicy {
-    return new TokenCredentialPolicy(nextPolicy, options, this);
+  getToken(_scopes: string | string[], _options?: GetTokenOptions): Promise<AccessToken | null> {
```
>getToken [](start = 2, length = 8)

Should add public and async markers for this method `public async getToken()`

daviwil replied:
Thanks!  Forgot the add the `public` marker. `async` shouldn't be necessary in this case since no `await` is used.

XiaoningLiu replied:
This method returns a Promise, so I assume it should be an async method. With async marker, we can skip `Promise.resolve` wrapper.

daviwil replied:
OK, I will change it to `async` and return the object directly.

XiaoningLiu commented on sdk/storage/storage-blob/src/credentials/RawTokenCredential.ts:
```diff
@@ -26,39 +23,36 @@ import { TokenCredentialPolicy } from "../policies/TokenCredentialPolicy";
+  getToken(_scopes: string | string[], _options?: GetTokenOptions): Promise<AccessToken | null> {
+    return Promise.resolve({
+      token: this.token,
+      expiresOnTimestamp: Date.now() + 2 * 60 * 1000 // 2 Minutes
```
>expiresOnTimestamp [](start = 6, length = 18)

We should assume token provided in `RawTokenCredential` will never expired, because it's up to customers refreshing token string value.

daviwil replied:
We've implemented token refreshing based on expiration time in `BearerTokenAuthenticationPolicy` so since it's possible for the user to change the `token`, we need a relatively short expiration so we aren't using an out of date token.

One possibility is that I could change that caching code to always ask for the current token when `expiresOnTimestamp` is set to `0`.

XiaoningLiu replied:
If am correct, `RawTokenCredential` is a simple wrapper for bearer token string. It needs manually refresh by customers upodating `token` property.

I'm not sure what happens after 2 mins timeout. Because `RawTokenCredential` doesn't define any token refresh logic. 

So what will hapen after 2mins timeout? Will underlayer just call `getToken` again? If so, this call can be optimized by setting up an infinite token..

---
In reply to: [295116310](https://github.com/Azure/azure-sdk-for-js/pull/3853#discussion_r295116310) [](ancestors = 295116310)

XiaoningLiu commented on sdk/storage/storage-blob/src/credentials/RawTokenCredential.ts:
```diff
@@ -26,39 +23,36 @@ import { TokenCredentialPolicy } from "../policies/TokenCredentialPolicy";
-  public create(nextPolicy: RequestPolicy, options: RequestPolicyOptions): TokenCredentialPolicy {
-    return new TokenCredentialPolicy(nextPolicy, options, this);
+  getToken(_scopes: string | string[], _options?: GetTokenOptions): Promise<AccessToken | null> {
+    return Promise.resolve({
```
After adding `async` marker, could return object directly without wrapping with Promise.

```javascript
return {
  token: xxx;
  expiresOnTimesteamp: xxx
};
```

XiaoningLiu commented on sdk/storage/storage-blob/src/credentials/RawTokenCredential.ts:
```diff
@@ -26,39 +23,36 @@ import { TokenCredentialPolicy } from "../policies/TokenCredentialPolicy";
    */
-  public create(nextPolicy: RequestPolicy, options: RequestPolicyOptions): TokenCredentialPolicy {
-    return new TokenCredentialPolicy(nextPolicy, options, this);
+  getToken(_scopes: string | string[], _options?: GetTokenOptions): Promise<AccessToken | null> {
```
>_scopes [](start = 11, length = 7)

Parameter should remove `_` prefix

daviwil replied:
I added those because the parameters are unused, we get compiler errors without the `_`.

XiaoningLiu commented on sdk/storage/storage-blob/src/credentials/RawTokenCredential.ts:
```diff
@@ -26,39 +23,36 @@ import { TokenCredentialPolicy } from "../policies/TokenCredentialPolicy";
+   * Retrieves the token stored in this RawTokenCredential.
+   * 
+   * @param _scopes Ignored since token is already known.
+   * @param _options Ignored since token is already known.
```
>param [](start = 6, length = 5)

JSDoc should include `@return` section for return values.

XiaoningLiu commented:
> export class TokenCredentialPolicy extends CredentialPolicy {

Should delete this file, as `RawTokenCredential` doesn't depend on this policy anymore

---
Refers to: sdk/storage/storage-blob/src/policies/TokenCredentialPolicy.ts:17 in 6313f3c. [](commit_id = 6313f3c9f30e411ca5c4dd325ea27153932cb3cb, deletion_comment = False)

XiaoningLiu commented:
> export class TokenCredentialPolicy extends CredentialPolicy {

Should remove this file as it's not necessary anymore.

---
Refers to: sdk/storage/storage-queue/src/policies/TokenCredentialPolicy.ts:18 in 6313f3c. [](commit_id = 6313f3c9f30e411ca5c4dd325ea27153932cb3cb, deletion_comment = False)

XiaoningLiu commented on sdk/storage/storage-queue/src/QueueClient.ts:
```diff
@@ -221,12 +225,16 @@ export class QueueClient extends StorageClient {
    *                     append a SAS if using AnonymousCredential, such as
    *                     "https://myaccount.queue.core.windows.net/myqueue?sasString".
-   * @param {Credential} credential Such as AnonymousCredential, SharedKeyCredential or TokenCredential.
+   * @param {Credential | TokenCredential } credential Such as AnonymousCredential, SharedKeyCredential or TokenCredential.
```
>TokenCredential [](start = 107, length = 15)

Such as AnonymousCredential, SharedKeyCredential RawTokenCredential or TokenCredentials retrieved from @azure/identity package

daviwil replied:
I'll update all of the comments across the various Client files

XiaoningLiu replied:
need updates in samples folder and readme.md

XiaoningLiu commented:
@jeremymeng Can you help open an issue to track the authentication merging between storage SDK and @azure/identity?

`Credential | TokenCredential' still feels not perfect, we can resolve this later before track2 GA.

jeremymeng commented:
I think it’s better to use non async since async method as overhead. Unless compiler knows how to optimize for this case…

Sent from my Windows 10 device

________________________________
From: David Wilson <notifications@github.com>
Sent: Tuesday, June 18, 2019 9:26:15 PM
To: Azure/azure-sdk-for-js
Cc: Jeremy Meng; Mention
Subject: Re: [Azure/azure-sdk-for-js] [Storage] Use @azure/core-http in @azure/storage-* libraries to support @azure/identity (#3853)


@daviwil commented on this pull request.

________________________________

In sdk/storage/storage-blob/src/credentials/RawTokenCredential.ts<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAzure%2Fazure-sdk-for-js%2Fpull%2F3853%23discussion_r295116448&data=02%7C01%7CJeremy.Meng%40microsoft.com%7Cb950f374f591495cdd1108d6f46e44d5%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636965151776016424&sdata=emTF9AG44ixzayHAjnTF0axhrxMgmT8fFW4UC0t2YQY%3D&reserved=0>:

>     */
-  public create(nextPolicy: RequestPolicy, options: RequestPolicyOptions): TokenCredentialPolicy {
-    return new TokenCredentialPolicy(nextPolicy, options, this);
+  getToken(_scopes: string | string[], _options?: GetTokenOptions): Promise<AccessToken | null> {


OK, I will change it to async and return the object directly.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAzure%2Fazure-sdk-for-js%2Fpull%2F3853%3Femail_source%3Dnotifications%26email_token%3DABZ3QX762ZZRS2ISNOTOCMTP3GYOPA5CNFSM4HYZ5VO2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB36RSPY%23discussion_r295116448&data=02%7C01%7CJeremy.Meng%40microsoft.com%7Cb950f374f591495cdd1108d6f46e44d5%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636965151776026416&sdata=l71DUUzZ5EY5elLxOFXGdd7xkzAPsh74zs3dfXYJAHQ%3D&reserved=0>, or mute the thread<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FABZ3QXZ7VK64DRTA7O7ODXTP3GYOPANCNFSM4HYZ5VOQ&data=02%7C01%7CJeremy.Meng%40microsoft.com%7Cb950f374f591495cdd1108d6f46e44d5%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636965151776026416&sdata=YpW4suryGEbtETMZpxgeloycljZBV%2Bknyq%2BzV6g0rVE%3D&reserved=0>.

daviwil commented:
@jeremymeng I was also concerned about that.  I just checked the compiler's output, there does seem to be a bit of overhead:

```
    RawTokenCredential.prototype.getToken = function (_scopes, _options) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                return [2 /*return*/, {
                        token: this.token,
                        expiresOnTimestamp: Date.now() + 2 * 60 * 1000 // 2 Minutes
                    }];
            });
        });
    };
```

I think I'll stick with no `async` and return `Promise.resolve`

jeremymeng commented:
> Can you help open an issue to track the authentication merging between storage SDK and @azure/identity?

Logged #3944

XiaoningLiu commented:
> @jeremymeng I was also concerned about that. I just checked the compiler's output, there does seem to be a bit of overhead:
> 
> ```
>     RawTokenCredential.prototype.getToken = function (_scopes, _options) {
>         return tslib_1.__awaiter(this, void 0, void 0, function () {
>             return tslib_1.__generator(this, function (_a) {
>                 return [2 /*return*/, {
>                         token: this.token,
>                         expiresOnTimestamp: Date.now() + 2 * 60 * 1000 // 2 Minutes
>                     }];
>             });
>         });
>     };
> ```
> 
> I think I'll stick with no `async` and return `Promise.resolve`

@jeremymeng How much performance drop for this overhead, or do we have some data? TS compile should resolve this, as async method is very widely used.

jeremymeng commented:
> @jeremymeng How much performance drop for this overhead

I think It's adding unnecessary overhead of the generated code. Async function will return a Promise too. Why not return the promise directly? It is still compatible with async/await.

daviwil commented:
While running the test suites for all 3 storage modules, I encountered a number of failures caused by generated model changes.  I believe there were changes to `autorest.typescript` more recently than the last time this code was generated which are causing problems with the XML serialization of some types.

I'll spend some time tomorrow getting those cleared up and then I'll send out the final samples and tests so that we can hopefully get this PR merged.

XiaoningLiu commented:
> > @jeremymeng How much performance drop for this overhead
> 
> I think It's adding unnecessary overhead of the generated code. Async function will return a Promise too. Why not return the promise directly? It is still compatible with async/await.

Well, most async functions can also be written as regular functions using Promises. But why introduced async marker? 

For this particular case, we have different starting points. I prefer less lines of and cleaner code in our TS code base (by removing the Promise.resolve wrapper), more developer friendly. Generated code optimization should be the work of TS compiler. Developers shouldn't care about how JS compiled code looks like. (We shipped the uglyfied code in npm package right?)

Another is about future consideration. Inside an async marked method, we can await other async methods and **throw async errors** directly, instead of using complex promise reject. If we need to modify this method in the future by someone others, we don't want the 

Last is about the alignment, in storage SDKs, all async methods are marked with async, just want no exception.

XiaoningLiu commented:
> While running the test suites for all 3 storage modules, I encountered a number of failures caused by generated model changes. I believe there were changes to `autorest.typescript` more recently than the last time this code was generated which are causing problems with the XML serialization of some types.
> 
> I'll spend some time tomorrow getting those cleared up and then I'll send out the final samples and tests so that we can hopefully get this PR merged.

Which autorest.typescript version are you using? I am working on features dev work for storage SDK using autorest.typescript@4.1.1, and it works well for me.

Besides the async change, others looks good to me. Hope this get resolved.

jeremymeng commented:
> Well, most async functions can also be written as regular functions using Promises. But why introduced async marker?

I don't propose changing async functions with `await` inside into promise resolve/reject.  My argument is for this specific simple case, between

```typescript
async getToken(): Promise<Token> {
  return token;
}
``` 

and

```typescript
getToken(): Promise<Token> {
  return Promise.resolve(token);
}
```

If we need to await other async functions inside this function in the future we can add `async`.

But I get your point about consistency.  @bterlson please help make a call :)

bterlson commented:
This is an area of ambiguity [in the guidelines](https://azuresdkspecs.z5.web.core.windows.net/TypeScriptSpec.html#ts-use-async-functions) but that's somewhat by design because it's a hard tradeoff as you've discovered :) It is true that using an async function here while targeting ES6 increases the code footprint of that function pretty substantially.

I'm personally in favor of the middle-ground approach I put in the guidelines - if you're going to return a thing that's a promise, just return it rather than make your function async. It's a bit of a code smell and might mean future refactorings, but it helps keep bundle sizes down.

I agree that the compiler should provide a simpler transformation that doesn't carry this weight (i.e. it emits promise.then code rather than a generator), but I think the team is reluctant because it's a huge change, isn't 100% compatible with native async function semantics, and we won't need it forever once async functions are supported everywhere. I continue to push for this however (/cc @DanielRosenwasser  😀)

daviwil commented:
Finally completed work on `storage-blob` and `storage-queue`.  I had to take the new `TokenCredential` back out of `storage-file` because it turns out [only `blob` and `queue` support AAD OAuth](https://docs.microsoft.com/en-us/azure/storage/common/storage-auth-aad) (bearer token) authentication.

I've taken the PR out of Draft state so feel free to take a final look over this whenever you have a chance.  I'll also review everything in more detail and make some tweaks if necessary.

DanielRosenwasser commented:
Can I get a sense of what code bloat you're getting when converting to ES2015? I know ES5 is pretty rough, but I'm having a hard time understanding where things go really wrong.

The following code

```ts
let x = {
  async getToken(): Promise<Token> {
    return token;
  }
};
```

is emitted as follows in ES2015

```js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
let x = {
    getToken() {
        return __awaiter(this, void 0, void 0, function* () {
            return token;
        });
    }
};
```

and when using `--importHelpers` with modules, that should look something like

```ts
import * as tslib from "tslib";

let x = {
    getToken() {
        return tslib.__awaiter(this, void 0, void 0, function* () {
            return token;
        });
    }
};
```

So can you give me an example of where things explode?

jeremymeng commented on sdk/storage/storage-blob/test/node/appendblobclient.spec.ts:
```diff
@@ -44,6 +46,17 @@ describe("AppendBlobClient Node.js only", () => {
     await newClient.download();
   });
 
+  it.only("can be created with a url and a TokenCredential", async () => {
```
Remove `.only`

daviwil replied:
Argh, thanks! I knew I was forgetting something

jeremymeng commented on sdk/storage/storage-file/src/DirectoryClient.ts:
```diff
@@ -181,8 +181,8 @@ export class DirectoryClient extends StorageClient {
    *                     Such as a directory named "mydir%", the URL should be "https://myaccount.file.core.windows.net/myshare/mydir%25".
-   * @param {Credential} [credential] Such as AnonymousCredential, SharedKeyCredential or TokenCredential.
-   *                                If not specified, AnonymousCredential is used.
+   * @param {Credential } [credential] Such as AnonymousCredential, SharedKeyCredential or TokenCredential.
```
nit extract space after `{Credential`

jeremymeng commented on sdk/storage/storage-file/src/FileServiceClient.ts:
```diff
@@ -174,8 +174,8 @@ export class FileServiceClient extends StorageClient {
    *                     if using AnonymousCredential, such as "https://myaccount.file.core.windows.net?sasString".
-   * @param {Credential} [credential] Such as AnonymousCredential, SharedKeyCredential or TokenCredential.
-   *                                If not specified, AnonymousCredential is used.
+   * @param {Credential } [credential] Such as AnonymousCredential, SharedKeyCredential or TokenCredential.
```
nit: extra space

jeremymeng commented:
I am fine with async function. I don't thin the discussion should block this PR.

daviwil commented:
OK, I've updated everything based on all of the review feedback so far.  A summary of the latest changes:

- Updated `BreakingChanges.md` in `storage-queue` and `storage-blob`
- Updated samples to rename `TokenCredential` to `RawTokenCredential`
- Fixed minor spacing and test issues Jeremy found
- Changed `RawTokenCredential.getToken` back to an `async` method

@XiaoningLiu, let me know if you have any final comments, otherwise I'll get this merged first thing Monday morning Pacific time.  Thanks!

XiaoningLiu commented on sdk/storage/storage-blob/samples/typescript/basic.ts:
```diff
@@ -19,7 +19,7 @@ async function main() {
   const sharedKeyCredential = new SharedKeyCredential(account, accountKey);
 
   // Use TokenCredential with OAuth token
```
>TokenCredential [](start = 9, length = 15)

typo `RawTokenCredential`

daviwil replied:
Thanks!  I'll fix the other instances too.

XiaoningLiu commented on sdk/storage/storage-blob/samples/typescript/azureAdAuth.ts:
```diff
@@ -0,0 +1,44 @@
+import { BlobServiceClient, SharedKeyCredential } from "../../src"; // Change to "@azure/storage-blob" in your package
+import { DefaultAzureCredential } from "@azure/identity";
+
+async function main() {
```
>main [](start = 15, length = 4)

Is "Ad" or "Aad"? is this file name `azureAdAuth.ts` correct?

daviwil replied:
I called it "azureAd" to refer to "Azure Active Directory", though I can see how having "aad" explicitly could make things clearer.  I'll ask someone about this and will rename these files after the PR gets merged if necessary.

XiaoningLiu commented on sdk/storage/storage-blob/samples/typescript/basic.ts:
```diff
@@ -19,7 +19,7 @@ async function main() {
   const sharedKeyCredential = new SharedKeyCredential(account, accountKey);
 
   // Use TokenCredential with OAuth token
```
>Use [](start = 5, length = 3)

Do we need to mention credentials from `@azure/identity` here?

daviwil replied:
Couldn't hurt, I'll update the comments to refer to it.

daviwil commented:
Thanks @XiaoningLiu, I've made updates to address your feedback.  Once CI goes green I will get this merged.

Thanks everyone!

HarshaNalluru commented:
> @HarshaNalluru @jeremymeng PR triggered checks skip all storage tests. Is this expected? Considering we added mocking test already, can we enable tests against PRs?

@XiaoningLiu,  
- "playing back the tests" is still being skipped in the feature/storage branch but is enabled in the master branch.
- I'll merge the master branch into the feature branch which would technically mean that feature branch will have the tests running in playback mode as part of PR validations.
- But the recordings should be updated in the feature branch since some of the tests are updated/different compared to the master branch.
